packages/reactivity/src/collectionHandlers.ts

```diff
@@ -66,8 +66,8 @@ function size(target: IterableCollections, isReadonly = false) {
   return Reflect.get(target, 'size', target)
 }
 
-function add(this: SetTypes, value: unknown) {
-  value = toRaw(value)
+function add(this: SetTypes, value: unknown, isShallow = false) {
```

Comment:
Adding a boolean flag to a function is generally something that sets off alarm bells for me. There's a lot of stuff been written on this topic, e.g. https://martinfowler.com/bliki/FlagArgument.html, but that isn't to say it's necessarily wrong.

When I first looked over this PR a few weeks ago, I felt like the boolean flag was justified here. But it kept nagging away at me, and I'm now wondering whether there's a better way.

First, I'd like to consider this example:

- [Playground - this PR](https://deploy-preview-8503--vue-sfc-playground.netlify.app/#eNp9kjFvwjAQhf/KyUtBSsPARgGprRjaoa2gW93BOAeYJnZkOwEpyn/v2REB1IrNd++783uWG/ZYlmldIZuwqZNWlR4c+qqcc62K0lgPDVgU0qsaE1Bu2Z+9WYoDtLCxpoA7WnHHNdfSaBdXwKyfG2g8wAr9YDg8AeJSbkBMwNsKoe2B9TWw/gPIa0BeAFw/fIkE1gnI73Rj7ELI3YAspSLLkuCNmOmoS0s5qfBYlLnwSBXANFM11Pc0OOPMrPegdBc27BhyFiGApoEgti3M5qE4v82A+kMS4rIRbaOJ6ai/gyXMOwqxUdt074ymp28Cypk0RalytO+lVxSSswlEJWgiz83hNfZC0OTUlzuUP//09+4Yepx9WHRoa+Ss17ywW/SdvFi94ZHOvViYrMqJviEu0Zm8Ch477KnSGdm+4KLbl/iBlN5+usXRo3anUMFoINvIc0af5/lG9LPdcTqOc1y3rP0FDqDpRA==)

It's a silly example, but it shows how the flag can be passed accidentally by client code. In that example, the index is being passed by `forEach` as the `isShallow` value.

Does this matter? Maybe not, but it motivated me to ponder alternative implementations.

I'm now wondering whether this might be a better way to implement it:
1. Remove the call to `toRaw` in `add` and `set`. Have them effectively be shallow by default.
2. Override the calls to `add` and `set` in `mutableInstrumentations` to include the `toRaw` call. Something like:

    ```js
    add(this: SetTypes, value: unknown) {
      return add.call(this, toRaw(value))
    },
    set(this: MapTypes, key: unknown, value: unknown) {
      return set.call(this, key, toRaw(value))
    },
    ```

This avoids the need for the boolean flag. I tried this locally and it seemed to pass the test cases.

Reply:
**Method 1** I don't think is advisable, it will lead to the contamination of the raw data, as shown in the code below, modifying the state through the raw data will cause the effect to be re-executed, and toRaw is used to avoid this

```js
 const m = new Map()
 const p1 = reactive(m)
 const p2 = reactive(new Map())
 p1.set('p2', p2)
 effect(() => {
 console.log(m.get('p2').size)
 })
 m.get('p2').set('foo', 1)
```

Reply:
I think you misunderstood what I meant. I wasn't proposing two separate methods for fixing the problem, those were two parts of a single method.

I've put together an example to show the changes I had in mind:

https://github.com/skirtles-code/vue-next/commit/073ab1c1e12b26cd0b7fc865f8438790966b0533

The calls to `toRaw` are still present, they're just moved.

Reply:
Haha, I'm sorry, I misunderstood

Reply:
I think adding the `isShallow` arg is fine in this case because it is consistent with how other functions in the same file are structured. The `isShallow` flag is used only for value unwrapping in `add` and `set`, but affects more things in other functions - in general, these handlers **have to** be aware of the context they are called in.